pytorch_lightning/metrics/functional/classification.py

```diff
@@ -998,12 +1000,37 @@ def iou(
         tensor(0.4914)
 
     """
+    num_classes = get_num_classes(pred=pred, target=target, num_classes=num_classes)
+
     tps, fps, tns, fns, sups = stat_scores_multiple_classes(pred, target, num_classes)
-    if remove_bg:
-        tps = tps[1:]
-        fps = fps[1:]
-        fns = fns[1:]
-    denom = fps + fns + tps
-    denom[denom == 0] = torch.tensor(FLOAT16_EPSILON).type_as(denom)
-    iou = tps / denom
-    return reduce(iou, reduction=reduction)
+
+    scores = torch.zeros(num_classes, device=pred.device, dtype=torch.float32)
+
+    for class_idx in range(num_classes):
+        # Skip this class if its index is being ignored.
+        if class_idx == ignore_index:
+            continue
+
+        tp = tps[class_idx]
+        fp = fps[class_idx]
+        fn = fns[class_idx]
+        sup = sups[class_idx]
+
+        # If this class is not present in either the target (no support) or the pred (no true or false positives), then
+        # use the not_present_score for this class.
+        if sup + tp + fp == 0:
+            scores[class_idx] = not_present_score
+            continue
+
+        denom = tp + fp + fn
+        score = tp.to(torch.float) / denom
```

Comment:
what happened to the line `denom[denom == 0] = torch.tensor(FLOAT16_EPSILON).type_as(denom)` to avoid zero division errors?

Reply:
Good thing to bring up. I don't believe it is necessary any more, but please correct me if my reasoning below is incorrect and we can add the `denom[denom == 0]` catch back in. I've added this in slightly shorter form as a comment in https://github.com/PyTorchLightning/pytorch-lightning/pull/3098/commits/d4ff9b413fb130ba4138e54bbee3fe9f6e907436

If `denom == 0`, then we can expand to get `tp + fp + fn == 0`. The above `not_present_score` conditional would have been triggered if `sup + tp + fp == 0`, in which case we would have hit the `continue` above and avoided `denom` computation altogether. So we will only fall through to `denom = tp + fp + fn` if `sup + tp + fp != 0`. We expand the latter with `sup == tp + fn` to get `(tp + fn) + tp + fp != 0` or `2*tp + fp + fn != 0`. If the latter is true, and given all vars are non-negative, we can conclude `tp + fn + fn != 0` if we get here, and therefore don't have to do the `denom[denom == 0]` check.

Does that sound correct?